cpp/src/strings/capitalize.cu

```diff
@@ -0,0 +1,356 @@
+         auto the_chr = *itr;
+         uint32_t code_point = detail::utf8_to_codepoint(the_chr);
+         detail::character_flags_table_type flag = code_point <= 0x00FFFF ? get_flags()[code_point] : 0;
+         if( (itr == d_str.begin()) ? IS_LOWER(flag) : IS_UPPER(flag) )
+           the_chr = detail::codepoint_to_utf8(get_case_table()[code_point]);
+         bytes += detail::bytes_in_char_utf8(the_chr);
+       }
+       return bytes;
+    }
+  };
+
+  //functor for executing string capitalization:
+  //
+  struct execute_capitalize: execute_base
+  {
+    execute_capitalize(column_device_view const d_column,
+                       character_flags_table_type const* d_flags,
+                       character_cases_table_type const* d_case_table,
+                       int32_t const* d_offsets,
+                       char* d_chars):
+      execute_base(d_column, d_flags, d_case_table, d_offsets, d_chars)
+    {
+    }
+    
+    __device__
+    int32_t operator()(size_type idx) override {
+      if( get_column().is_null(idx) )
+        return 0; // null string
+      
+      string_view d_str = get_column().template element<string_view>(idx);
+      char* d_buffer = get_chars() + get_offsets()[idx];
+      
+      for( auto itr = d_str.begin(); itr != d_str.end(); ++itr ) {
+        auto the_chr = *itr;
+        uint32_t code_point = detail::utf8_to_codepoint(the_chr);
+        detail::character_flags_table_type flag = code_point <= 0x00FFFF ? get_flags()[code_point] : 0;
+
+        if( (itr == d_str.begin()) ? IS_LOWER(flag) : IS_UPPER(flag) )
+          the_chr = detail::codepoint_to_utf8(get_case_table()[code_point]);
+        d_buffer += detail::from_char_utf8(the_chr, d_buffer);
+      }
+      return 0;
+    }
+  };
+
+
+  //functor for probing string title-ization
+  //requirements:
+  //
+  struct probe_title: probe_base
+  {
+    probe_title(column_device_view const d_column,
+                     character_flags_table_type const* d_flags,
+                     character_cases_table_type const* d_case_table):
+      probe_base(d_column, d_flags, d_case_table)
+    {  
+    }
+    
+     __device__
+     int32_t operator()(size_type idx) const override {
+       if( get_column().is_null(idx) )
+         return 0; // null string
+      
+       string_view d_str = get_column().template element<string_view>(idx);
+       int32_t bytes = 0;
+
+       bool bcapnext = true;
+       for( auto itr = d_str.begin(); itr != d_str.end(); ++itr ) {
+         auto the_chr = *itr;
+         uint32_t code_point = detail::utf8_to_codepoint(the_chr);
+         detail::character_flags_table_type flag = code_point <= 0x00FFFF ? get_flags()[code_point] : 0;
+
+         if( !IS_ALPHA(flag) ) {
+           bcapnext = true;
+           bytes += detail::bytes_in_char_utf8(the_chr);
+         }
+         else if( (bcapnext && IS_LOWER(flag)) || (!bcapnext && IS_UPPER(flag)) ) {
+           bcapnext = false;
```

Comment:
The logic here needs to match logic for the execute pass.
Here the `bcapnext` is not being set to `false` in the else statement.
This means it will calculate the incorrect size for the chars column and possibly overrun the output buffer.
Add the following string to the Title testcase to see the error: `"SȺȺnich"`

It would be good if the probe and execute functions could be factored together to avoid this kind of coding error.

Reply:
Fixed `bcapnext` issue.